File: backend/routers/documents.py
"""Document upload and AI parsing endpoints."""
import os
import json
from fastapi import APIRouter, Depends, HTTPException, UploadFile, File
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select

from database import get_db
from models.campaign import Project, AdSet, Ad, ProjectStatus
from schemas.document import ParsedStructure
from pydantic import BaseModel
from typing import Optional, Dict, Any
from services.claude_parser import parse_document, validate_parsed_completeness
from utils.file_processing import extract_text
from config import get_settings

router = APIRouter(prefix="/api/documents", tags=["documents"])


@router.post("/{project_id}/upload")
async def upload_document(
    project_id: int,
    file: UploadFile = File(...),
    db: AsyncSession = Depends(get_db),
):
    """Upload a campaign structure document (PDF, DOCX, TXT)."""
    result = await db.execute(select(Project).where(Project.id == project_id))
    project = result.scalar_one_or_none()
    if not project:
        raise HTTPException(404, "Project not found")

    # Save file temporarily
    settings = get_settings()
    upload_dir = os.path.join(settings.storage_path, str(project_id), "docs")
    os.makedirs(upload_dir, exist_ok=True)
    file_path = os.path.join(upload_dir, file.filename)

    with open(file_path, "wb") as f:
        content = await file.read()
        f.write(content)

    # Extract text
    try:
        text = extract_text(file_path)
    except ValueError as e:
        raise HTTPException(400, str(e))

    project.raw_document_text = text
    await db.commit()

    return {"status": "uploaded", "filename": file.filename, "text_length": len(text)}


@router.post("/{project_id}/parse")
async def parse_structure(project_id: int, db: AsyncSession = Depends(get_db)):
    """Parse the uploaded document with AI to extract campaign structure."""
    result = await db.execute(select(Project).where(Project.id == project_id))
    project = result.scalar_one_or_none()
    if not project:
        raise HTTPException(404, "Project not found")
    if not project.raw_document_text:
        raise HTTPException(400, "No document uploaded. Upload a document first.")

    try:
        structure = await parse_document(project.raw_document_text)
    except Exception as e:
        raise HTTPException(500, f"AI parsing failed: {e}")

    # Save parsed structure JSON
    project.parsed_structure_json = structure.model_dump_json()

    # Update project from parsed campaign-level fields
    if structure.campaign.name:
        project.name = structure.campaign.name
    if structure.campaign.objective:
        project.campaign_objective = structure.campaign.objective
    if structure.campaign.budget_type:
        from models.campaign import BudgetType
        project.budget_type = BudgetType(structure.campaign.budget_type)
    if structure.campaign.daily_budget is not None:
        project.daily_budget = structure.campaign.daily_budget
    if structure.campaign.destination_url:
        project.destination_url = structure.campaign.destination_url

    # Create AdSet and Ad records from parsed structure
    # First, clear existing ad sets for this project
    existing_sets = await db.execute(select(AdSet).where(AdSet.project_id == project_id))
    for adset in existing_sets.scalars().all():
        await db.delete(adset)
    await db.flush()

    for parsed_adset in structure.ad_sets:
        adset = AdSet(
            project_id=project_id,
            name=parsed_adset.name,
            targeting_json=parsed_adset.targeting.model_dump_json(),
            placements_json=(
                json.dumps(parsed_adset.placements)
                if isinstance(parsed_adset.placements, dict)
                else None
            ),
            budget=parsed_adset.budget_daily,
            bid_strategy=parsed_adset.bid_strategy,
            optimization_goal=parsed_adset.optimization_goal,
        )
        db.add(adset)
        await db.flush()

        for parsed_ad in parsed_adset.ads:
            ad = Ad(
                ad_set_id=adset.id,
                name=parsed_ad.name,
                creative_ref=parsed_ad.creative_ref,
                headline=parsed_ad.headline,
                primary_text=parsed_ad.primary_text,
                description=parsed_ad.description,
                cta=parsed_ad.cta,
                url=parsed_ad.url,
            )
            db.add(ad)

    project.status = ProjectStatus.parsed
    await db.commit()

    # Check for missing required fields
    missing_fields = validate_parsed_completeness(structure)

    return {
        "status": "parsed",
        "structure": structure.model_dump(),
        "missing_fields": missing_fields,
        "ad_sets_count": len(structure.ad_sets),
        "total_ads": sum(len(s.ads) for s in structure.ad_sets),
    }


@router.get("/{project_id}/parsed")
async def get_parsed_structure(project_id: int, db: AsyncSession = Depends(get_db)):
    """Get the parsed campaign structure."""
    result = await db.execute(select(Project).where(Project.id == project_id))
    project = result.scalar_one_or_none()
    if not project:
        raise HTTPException(404, "Project not found")
    if not project.parsed_structure_json:
        raise HTTPException(400, "No parsed structure available")

    return json.loads(project.parsed_structure_json)


@router.put("/{project_id}/parsed")
async def update_parsed_structure(
    project_id: int,
    structure: ParsedStructure,
    db: AsyncSession = Depends(get_db),
):
    """Manually edit the parsed structure."""
    result = await db.execute(select(Project).where(Project.id == project_id))
    project = result.scalar_one_or_none()
    if not project:
        raise HTTPException(404, "Project not found")

    project.parsed_structure_json = structure.model_dump_json()
    await db.commit()

    return {"status": "updated"}


class CompleteFieldsRequest(BaseModel):
    campaign_name: Optional[str] = None
    campaign_objective: Optional[str] = None
    budget_type: Optional[str] = None
    daily_budget: Optional[float] = None
    destination_url: Optional[str] = None


@router.post("/{project_id}/complete-fields")
async def complete_missing_fields(
    project_id: int,
    data: CompleteFieldsRequest,
    db: AsyncSession = Depends(get_db),
):
    """Fill in missing required fields that weren't in the strategy document."""
    result = await db.execute(select(Project).where(Project.id == project_id))
    project = result.scalar_one_or_none()
    if not project:
        raise HTTPException(404, "Project not found")

    if data.campaign_name:
        project.name = data.campaign_name
    if data.campaign_objective:
        project.campaign_objective = data.campaign_objective
    if data.budget_type:
        from models.campaign import BudgetType
        project.budget_type = BudgetType(data.budget_type)
    if data.daily_budget is not None:
        project.daily_budget = data.daily_budget
    if data.destination_url:
        project.destination_url = data.destination_url

    # Also update the parsed_structure_json if it exists
    if project.parsed_structure_json:
        parsed = json.loads(project.parsed_structure_json)
        campaign = parsed.get("campaign", {})
        if data.campaign_name:
            campaign["name"] = data.campaign_name
        if data.campaign_objective:
            campaign["objective"] = data.campaign_objective
        if data.budget_type:
            campaign["budget_type"] = data.budget_type
        if data.daily_budget is not None:
            campaign["daily_budget"] = data.daily_budget
        if data.destination_url:
            campaign["destination_url"] = data.destination_url
        parsed["campaign"] = campaign
        project.parsed_structure_json = json.dumps(parsed)

    await db.commit()

    # Verify completeness
    remaining = []
    if not project.name or not project.name.strip():
        remaining.append("campaign_name")
    if not project.campaign_objective:
        remaining.append("campaign_objective")
    if not project.budget_type:
        remaining.append("budget_type")
    if project.daily_budget is None:
        remaining.append("daily_budget")
    if not project.destination_url:
        remaining.append("destination_url")

    return {
        "status": "completed" if not remaining else "incomplete",
        "missing_fields": remaining,
    }

File: api/lib/routes/documents.ts
import { Router, Request, Response } from "express";
import { eq } from "drizzle-orm";
import multer from "multer";
import fs from "fs";
import path from "path";
import { db } from "../db/client.js";
import { projects, adSets, ads } from "../db/schema.js";
import { parseDocument, validateParsedCompleteness } from "../services/claude-parser.js";
import { extractText } from "../services/file-processing.js";

const upload = multer({ dest: "/tmp/uploads" });
const router = Router();

// ── POST /:id/upload ────────────────────────────────────────────────────────

router.post(
  "/:id/upload",
  upload.single("file"),
  async (req: Request, res: Response) => {
    try {
      const projectId = parseInt(req.params.id, 10);

      const projectRows = await db
        .select()
        .from(projects)
        .where(eq(projects.id, projectId))
        .limit(1);
      const project = projectRows[0];
      if (!project) {
        return res.status(404).json({ detail: "Project not found" });
      }

      const file = req.file;
      if (!file) {
        return res.status(400).json({ detail: "No file uploaded" });
      }

      // Move file to a stable location
      const uploadDir = path.join("/tmp/uploads", String(projectId), "docs");
      fs.mkdirSync(uploadDir, { recursive: true });
      const destPath = path.join(uploadDir, file.originalname);
      fs.renameSync(file.path, destPath);

      // Extract text from file
      let text: string;
      try {
        text = await extractText(destPath);
      } catch (e: any) {
        return res.status(400).json({ detail: e.message });
      }

      await db
        .update(projects)
        .set({ rawDocumentText: text })
        .where(eq(projects.id, projectId));

      return res.json({
        status: "uploaded",
        filename: file.originalname,
        text_length: text.length,
      });
    } catch (e: any) {
      return res.status(e.status ?? 500).json({ detail: e.message ?? "Internal server error" });
    }
  },
);

// ── POST /:id/parse ─────────────────────────────────────────────────────────

router.post("/:id/parse", async (req: Request, res: Response) => {
  try {
    const projectId = parseInt(req.params.id, 10);

    const projectRows = await db
      .select()
      .from(projects)
      .where(eq(projects.id, projectId))
      .limit(1);
    const project = projectRows[0];
    if (!project) {
      return res.status(404).json({ detail: "Project not found" });
    }
    if (!project.rawDocumentText) {
      return res.status(400).json({ detail: "No document uploaded. Upload a document first." });
    }

    let structure: any;
    try {
      structure = await parseDocument(project.rawDocumentText);
    } catch (e: any) {
      return res.status(500).json({ detail: `AI parsing failed: ${e.message}` });
    }

    // Update project fields from parsed campaign-level data
    const campaignUpdates: Record<string, unknown> = {
      parsedStructureJson: JSON.stringify(structure),
      status: "parsed",
    };
    if (structure.campaign?.name) {
      campaignUpdates.name = structure.campaign.name;
    }
    if (structure.campaign?.objective) {
      campaignUpdates.campaignObjective = structure.campaign.objective;
    }
    if (structure.campaign?.budget_type) {
      campaignUpdates.budgetType = structure.campaign.budget_type;
    }
    if (structure.campaign?.daily_budget != null) {
      campaignUpdates.dailyBudget = structure.campaign.daily_budget;
    }
    if (structure.campaign?.destination_url) {
      campaignUpdates.destinationUrl = structure.campaign.destination_url;
    }

    await db.update(projects).set(campaignUpdates).where(eq(projects.id, projectId));

    // Clear existing ad sets for this project
    const existingAdSets = await db
      .select()
      .from(adSets)
      .where(eq(adSets.projectId, projectId));
    for (const adSet of existingAdSets) {
      await db.delete(ads).where(eq(ads.adSetId, adSet.id));
    }
    await db.delete(adSets).where(eq(adSets.projectId, projectId));

    // Create AdSet and Ad records from parsed structure
    let totalAds = 0;
    for (const parsedAdSet of structure.ad_sets ?? []) {
      const adSetResult = await db
        .insert(adSets)
        .values({
          projectId,
          name: parsedAdSet.name,
          targetingJson:
            typeof parsedAdSet.targeting === "object"
              ? JSON.stringify(parsedAdSet.targeting)
              : parsedAdSet.targeting ?? null,
          placementsJson:
            typeof parsedAdSet.placements === "object"
              ? JSON.stringify(parsedAdSet.placements)
              : parsedAdSet.placements === "automatic"
                ? null
                : null,
          budget: parsedAdSet.budget_daily ?? null,
          bidStrategy: parsedAdSet.bid_strategy ?? "LOWEST_COST_WITHOUT_CAP",
          optimizationGoal: parsedAdSet.optimization_goal ?? "OFFSITE_CONVERSIONS",
        })
        .returning();
      const adSet = adSetResult[0];

      for (const parsedAd of parsedAdSet.ads ?? []) {
        await db.insert(ads).values({
          adSetId: adSet.id,
          name: parsedAd.name,
          creativeRef: parsedAd.creative_ref ?? null,
          headline: parsedAd.headline ?? null,
          primaryText: parsedAd.primary_text ?? null,
          description: parsedAd.description ?? null,
          cta: parsedAd.cta ?? "SHOP_NOW",
          url: parsedAd.url ?? null,
        });
        totalAds++;
      }
    }

    // Check for missing required fields
    const missingFields = validateParsedCompleteness(structure);

    return res.json({
      status: "parsed",
      structure,
      missing_fields: missingFields,
      ad_sets_count: (structure.ad_sets ?? []).length,
      total_ads: totalAds,
    });
  } catch (e: any) {
    return res.status(e.status ?? 500).json({ detail: e.message ?? "Internal server error" });
  }
});

// ── GET /:id/parsed ─────────────────────────────────────────────────────────

router.get("/:id/parsed", async (req: Request, res: Response) => {
  try {
    const projectId = parseInt(req.params.id, 10);

    const projectRows = await db
      .select()
      .from(projects)
      .where(eq(projects.id, projectId))
      .limit(1);
    const project = projectRows[0];
    if (!project) {
      return res.status(404).json({ detail: "Project not found" });
    }
    if (!project.parsedStructureJson) {
      return res.status(400).json({ detail: "No parsed structure available" });
    }

    return res.json(JSON.parse(project.parsedStructureJson));
  } catch (e: any) {
    return res.status(e.status ?? 500).json({ detail: e.message ?? "Internal server error" });
  }
});

// ── PUT /:id/parsed ─────────────────────────────────────────────────────────

router.put("/:id/parsed", async (req: Request, res: Response) => {
  try {
    const projectId = parseInt(req.params.id, 10);

    const projectRows = await db
      .select()
      .from(projects)
      .where(eq(projects.id, projectId))
      .limit(1);
    if (!projectRows[0]) {
      return res.status(404).json({ detail: "Project not found" });
    }

    const structure = req.body;
    await db
      .update(projects)
      .set({ parsedStructureJson: JSON.stringify(structure) })
      .where(eq(projects.id, projectId));

    return res.json({ status: "updated" });
  } catch (e: any) {
    return res.status(e.status ?? 500).json({ detail: e.message ?? "Internal server error" });
  }
});

// ── POST /:id/complete-fields ───────────────────────────────────────────────

router.post("/:id/complete-fields", async (req: Request, res: Response) => {
  try {
    const projectId = parseInt(req.params.id, 10);

    const projectRows = await db
      .select()
      .from(projects)
      .where(eq(projects.id, projectId))
      .limit(1);
    const project = projectRows[0];
    if (!project) {
      return res.status(404).json({ detail: "Project not found" });
    }

    const {
      campaign_name,
      campaign_objective,
      budget_type,
      daily_budget,
      destination_url,
    } = req.body;

    const updates: Record<string, unknown> = {};
    if (campaign_name) updates.name = campaign_name;
    if (campaign_objective) updates.campaignObjective = campaign_objective;
    if (budget_type) updates.budgetType = budget_type;
    if (daily_budget != null) updates.dailyBudget = daily_budget;
    if (destination_url) updates.destinationUrl = destination_url;

    // Also update the parsed_structure_json if it exists
    if (project.parsedStructureJson) {
      const parsed = JSON.parse(project.parsedStructureJson);
      const campaign = parsed.campaign ?? {};
      if (campaign_name) campaign.name = campaign_name;
      if (campaign_objective) campaign.objective = campaign_objective;
      if (budget_type) campaign.budget_type = budget_type;
      if (daily_budget != null) campaign.daily_budget = daily_budget;
      if (destination_url) campaign.destination_url = destination_url;
      parsed.campaign = campaign;
      updates.parsedStructureJson = JSON.stringify(parsed);
    }

    if (Object.keys(updates).length > 0) {
      await db.update(projects).set(updates).where(eq(projects.id, projectId));
    }

    // Reload to verify completeness
    const reloaded = (
      await db.select().from(projects).where(eq(projects.id, projectId)).limit(1)
    )[0];

    const remaining: string[] = [];
    if (!reloaded.name || !reloaded.name.trim()) remaining.push("campaign_name");
    if (!reloaded.campaignObjective) remaining.push("campaign_objective");
    if (!reloaded.budgetType) remaining.push("budget_type");
    if (reloaded.dailyBudget == null) remaining.push("daily_budget");
    if (!reloaded.destinationUrl) remaining.push("destination_url");

    return res.json({
      status: remaining.length === 0 ? "completed" : "incomplete",
      missing_fields: remaining,
    });
  } catch (e: any) {
    return res.status(e.status ?? 500).json({ detail: e.message ?? "Internal server error" });
  }
});

export default router;
